nn.Transformer

zhangguanheng66 opened the pull request:
Create a PR for comments. The model is still WIP but I want to have some feedbacks before moving too far. The transformer model depends on several modules, like MultiheadAttention (landed).

Transformer is implemented based on the paper (https://papers.nips.cc/paper/7181-attention-is-all-you-need.pdf). Users have the flexibility to build a transformer with self-defined and/or built-in components (i.e encoder, decoder, encoder_layer, decoder_layer). Users could use Transformer class to build a standard transformer model and modify sub-layers as needed.

Add a few unit tests for the transformer module, as follow:
TestNN.test_Transformer_cell
TestNN.test_transformerencoderlayer
TestNN.test_transformerdecoderlayer
TestNN.test_transformer_args_check
TestScript.test_scriptmodule_transformer_cuda

There is another demonstration example for applying transformer module on the word language problem. https://github.com/pytorch/examples/pull/555

Commit:
A new PR for transformer model.

SsnL commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,290 @@
+from .dropout import Dropout
+from .linear import Linear
+from .normalization import LayerNorm
+
```
this likely breaks lint

SsnL commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,290 @@
+from .linear import Linear
+from .normalization import LayerNorm
+
+class TransformerBase(Module):
```
do you really need a base when there is only one subclass?

zhangguanheng66 replied:
This base class just lays out the encoder-decoder architecture.

cpuhrsch replied:
Let's remove this until we need an abstraction. Just like @SsnL mentioned it's only used in a single class. It also doesn't do more than add those two fields.

SsnL commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,290 @@
+            Unmasked positions are filled with float(0.0).
+        """
+        attn_shape = (sz, sz)
+        mask = np.triu(np.ones(attn_shape), k=1).astype('uint8')
```
we don't have hard dependency on numpy

SsnL replied:
this function also seems inefficient and unused everywhere

zhangguanheng66 replied:
Will fix it.

SsnL commented on torch/nn/modules/__init__.py:
```diff
@@ -27,6 +27,8 @@
 from .distance import PairwiseDistance, CosineSimilarity
 from .fold import Fold, Unfold
 from .adaptive import AdaptiveLogSoftmaxWithLoss
+from .transformer import TransformerBase, TransformerEncoder, TransformerDecoder, \
```
Why are the modules other than `Transformer` exposed?

zhangguanheng66 replied:
sometime users may use parts of the transformer model, so the encoder/decoder  modules are exposed. I agree TransformerBase should not be exposed.

SsnL commented:
Nice progress. But I have several comments and I also don't see the test and doc entries.

zhangguanheng66 commented:
> Nice progress. But I have several comments and I also don't see the test and doc entries.

We haven't decided a unit test for the transformer model. I will add the doc entries.

Commit:
Avoid expose TransformerBase
Add transformer doc string.

Commit:
Merge branch 'master' into transform_new_pr

cpuhrsch commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,290 @@
+from .module import Module
+from .activation import MultiheadAttention
+from .container import ModuleList
+import numpy as np
```
Remove this once you've deleted the numpy usage. We don't have a hard dependency on numpy.

cpuhrsch commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,290 @@
+                 num_decoder_layers=6, dim_feedforward=2048, dropout=0.1):
+        encoder_layer = TransformerEncoderLayer(d_model, nhead, dim_feedforward, dropout)
+        encoder_norm = LayerNorm(d_model) 
+        encoder = TransformerEncoder(encoder_layer, num_encoder_layers, encoder_norm)
```
We should make encoder and decoder keyword arguments and create this if they're set to None. This allows users to pass in custom code.

cpuhrsch commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,290 @@
+        self.dropout2 = Dropout(dropout)
+        self.dropout3 = Dropout(dropout)
+
+    def forward(self, tgt, memory, tgt_mask=None, memory_mask=None):
```
For all the forward functions, please move the compute into [functional.py](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1391). We separate the computations from the state. You can call the function you're adding something like "transformer_decoder_layer". Also, try annotating it with "@weak_script" just like the example I'm referencing above.

zhangguanheng66 replied:
Mitigate them soon.

cpuhrsch commented:
Looking much better! Just as @SsnL mentioned, please add unit tests.

Commit:
Remove TransformerBase class from transformer.py.
remove the dependency of numpy library in Transformer.py.
Add costom_encoder and costom_decoder to Transformer class.

Commit:
Add a unit test for Transformer -- test_Transformer_cell
Add a unit test for Transformer -- test_transformer_args_check
Add a unit test for Transformer -- TestNN.test_transformerencoderlayer
Add a unit test for Transformer -- TestNN.test_transformerencoderlayer
Add fixed numerical results.
Update TestNN.test_transformer_args_check to include src_mask, tgt_mask, memory_mask arguments

cpuhrsch commented on test/test_nn.py:
```diff
@@ -5228,6 +5228,129 @@ def test_LSTM_cell(self):
+            p.data.copy_(x)
+
+        # deterministic input
+        encoder_input = torch.Tensor([[[0.7462, 0.6653, 0.5679, 0.4891],
```
I think you can add more with even smaller and simpler inputs

cpuhrsch replied:
smaller here means "smaller tensors" and simpler means "integers"

Commit:
Add additional small tensor tests in test_transformerdecoderlayer and test_transformerencoderlayer.

Commit:
Fix a few lint errors.

cpuhrsch commented:
Pinging @srush @kyunghyuncho @myleott @glample for additional feature requests if any.

cpuhrsch commented:
Pinging @mansimov @jasonleeinf  @myleott @fmassa for additional feature requests / review.

cpuhrsch commented:
Pinging @stephenroller @douwekiela for additional feature requests / review.

zhangguanheng66 commented:
We also have a PR for applying torch.nn.Transformer in word language problem (https://github.com/pytorch/examples/pull/555). @myleott

stephenroller commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,287 @@
+        Shape:
+            see the docs in Transformer class.
+        """
+        src2 = self.norm1(src)
```
You should note that there's some debate about whether these go.

zhangguanheng66 replied:
One option is to check before calling the norm function.

cpuhrsch replied:
@stephenroller - do you mean there's debate around what kind of normalization to use or any at all?

@zhangguanheng66 - the docs should mention that we use norm, but I don't think we should make that optional / replaceable for now (unless we get many more comments).

cpuhrsch replied:
It might be easier to reimplement the EncoderLayer than making this one more complicated. But maybe I'm wrong.

zhangguanheng66 replied:
We should finally write a post to review the nn.Transformer module and clearly lay out the architecture, including norm.

stephenroller replied:
sorry I wrote whether but I meant "where".

Some variations put them before the selfattn/relu. Some put them after.

douwekiela replied:
Maybe we should benchmark a bunch of alternatives and pick the best one?

stephenroller replied:
Myle and Guillaume have for a bunch of things already. I’ve found it makes a *small* difference in my experiments but usually had too many confounding variables to make a clear recommendation.

cpuhrsch replied:
I suggest we do that in a separate PR since it's not a necessary feature to get the basic featureset.

stephenroller replied:
Just to actually ground it in code:

https://github.com/pytorch/fairseq/blob/dffb167449cf0e3ead73b4a67e84f222311b0ff5/fairseq/models/transformer.py#L532-L557

cpuhrsch replied:
Ok, let's standardize on ["Attention is all you need"](https://arxiv.org/abs/1706.03762) for now and mention that in the docs. Later on we can extend the behavior to match the implementation of other Transformer implementations with corresponding paper references.

cpuhrsch replied:
Also see [a comment in fairseq mentioning this](https://github.com/pytorch/fairseq/blob/dffb167449cf0e3ead73b4a67e84f222311b0ff5/fairseq/models/transformer.py#L570).

Commit:
Merge remote-tracking branch 'origin/master' into transform_new_pr

Commit:
Fix a typo in MultiheadAttention doc.

Commit:
Add an asserterror check back to the __init__ function of MultiheadAttention module.

Commit:
Merge branch 'mulihead_attn_doc' into transform_new_pr

Commit:
Merge remote-tracking branch 'origin/master' into transform_new_pr

Commit:
Split forword functions in encoder/decoder layers.

cpuhrsch commented:
cc @BIT-silence @kimishpatel

cpuhrsch commented:
Travis lint failures are relevant and need to be resolved (lines too long).

cpuhrsch commented on torch/nn/functional.py:
```diff
@@ -3262,3 +3263,147 @@ def _in_proj_v(weight, bias, embed_dim, value):
     return attn_output, attn_output_weights
+
+@weak_script
+def transformer_encoder_layer_forward(src,                                  # type: Tensor
```
Add to the docs that this is an unstable API, because we'll need to add support for slightly different standard transformers (e.g. BERT) and this is based on "All you need is attention" (which we should also cite and reference).

cpuhrsch commented on torch/nn/functional.py:
```diff
@@ -3262,3 +3263,147 @@ def _in_proj_v(weight, bias, embed_dim, value):
+    return src
+
+@weak_script
+def transformer_decoder_layer_forward(tgt,                                  # type: Tensor
```
Add to the docs that this is an unstable API, because we'll need to add support for slightly different standard transformers (e.g. BERT) and this is based on "All you need is attention" (which we should also cite and reference).

Commit:
Revert an early commit (i.e. forward functions in nn.functional.py for transformer module)

Commit:
Merge remote-tracking branch 'origin/master' into transformer

Commit:
Merge remote-tracking branch 'upstream/master' into transformer

cpuhrsch commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,288 @@
+
+class Transformer(Module):
+    r"""A transformer model. User is able to modified the attributes as needed. The architechture
+        is based on the paper "Attention Is All You Need".
```
Add full citation

zhangguanheng66 replied:
added.

cpuhrsch commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,288 @@
+        output = self.decoder(tgt, memory, tgt_mask, memory_mask)
+        return output
+
+    def generate_square_subsequent_mask(self, sz):
```
Doesn't seem to be used by models - maybe we want to remove it?

cpuhrsch commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,288 @@
+
+        return output
+
+class TransformerEncoderLayer(Module):
```
Add in details on standardization choice from discussion below.

Commit:
Add a jit unit test (i.e. test_scriptmodule_transformer_cuda) for transformer.

Commit:
Add full citation for transformer.

Commit:
Fix two lint errors.

cpuhrsch commented on test/test_jit.py:
```diff
@@ -12572,6 +12572,45 @@ def forward(self, q, k, v):
         # print(torch.allclose(jit_out, py_out, atol=5e-4, rtol=1e-4))
         self.assertTrue(torch.allclose(jit_out, py_out, atol=5e-4, rtol=1e-4))
 
+    @unittest.skipIf(not RUN_CUDA, "no CUDA")
```
Why no CUDA?

cpuhrsch commented on torch/nn/modules/transformer.py:
```diff
@@ -235,6 +240,8 @@ def forward(self, src, src_mask=None):
 
 class TransformerDecoderLayer(Module):
     r"""TransformerDecoderLayer is made up of self-attn, multi-head-attn and feedforward network.
+        This standard decoder layer is based on the paper "Attention Is All You Need".
```
Reference the note with the full citation for completion.

zhangguanheng66 replied:
added

Commit:
Add additional ref.

Commit:
Merge branch 'master' into transform_new_pr

zhangguanheng66 commented:
fix https://github.com/pytorch/pytorch/issues/10459

Commit:
Merge branch 'master' into transform_new_pr

DNGros commented on torch/nn/modules/transformer.py:
```diff
@@ -0,0 +1,303 @@
+
+    def __init__(self, encoder_layer, num_layers, norm=None):
+        super(TransformerEncoder, self).__init__()
+        self.layers = _get_clones(encoder_layer, num_layers)
```
Is it intended behavior to have weight sharing? In a standard transformer (like from Attention is All You Need), I don't believe the weights are shared. I understand there has been followup work showing weight sharing can be beneficial in some cases, but if this is intended behavior, it might be useful to specify this difference from the paper's implementation in docs after the Vaswani et al. citation and cite the followup work.

-----

As a broader point: 
This seems to get at the fact that a Transformer is a rather high level component, without really consensus yet on its architecture with a lot decisions likely domain specific (weight sharing?, beam search decoder?, convolutions between layers?, output the intermediate attentions? hierarchy / grouping components of the task?, sparsity?... and many more ideas that are being rapidly published) 

I don't mean to try and start needless debates, or to at all imply this Transformer code is not useful, but I will add in my thoughts that from my limited perspective that maybe Transformers might better belong in the contrib module or in the docs as a example that people can modify to meet their needs. The torch.nn.Modules module seems (at least currently) to be for more foundational components that can be composed into larger models, not full model architectures themselves.  

There has already been fairly extensive discussion on this in #10459 , and it seemed like the consensus there was to focus on things like MultiHeadedAttention or PositionalEncoding for core and keep full architectures separate in the codebase.

DNGros replied:
As an addendum to that second part: Thinking about this more I could see how a TransformerEncoder (or like StackedSelfAttention or RecurrentSelfAttention) could maybe be considered a primitive component which could lend itself to maybe eventually being low level optimized and composed into novel things (though not really that much more primitive than something like a ResNet-block). However, when it starts seeming particularly high level, and full-archetecture-y is the inclusion of a prepackaged encoder and decoder Transformer for seq2seq, which, without a lot of additional components, seems likely would not meet all needs and could not be easily adapted/composed. 
I don't know, and I don't really have any reason to hold strong opinions here. Feel free to dismiss this second part without real justification...

zhangguanheng66 replied:
@DNGros Thanks for the comments. To your first question, it's not supposed to share the weights between layers. Actually, if you take a quick look at _get_clones function, it calls a "deepcopy" function.

For your second comment, we do realize there are many ongoing discussions about the transformer models and several variants in different domains. We want to provide a baseline for the research community and startups when people don't want to code from scratch. This module could help people try some preliminary ideas for the fast delivery. For some advanced users, we expect them to develop a specific transformer model, and they could possibly use this module as a reference or benchmark case. 

As you suggest, we try to make the model highly "modularized". People can use nn.Tranformer, nn.TransformerEncoder, or EVEN nn.TransformerEncoderLayer, as needed. There is actually a word language example (https://github.com/pytorch/examples/pull/555) where we use nn.TransformerEncoder as the seq2seq model.

In the future, if we see a variant of the transformer model requested widely by the community, we will continuously implement them in our framework. I see a baseline model could benefit more users as we optimize the module performance in the future.

DNGros replied:
@zhangguanheng66 Ok, sorry. I missed that.

Thanks for clarifying!

Commit:
Move the layer_norm to the end of each sub-layers.

cpuhrsch reviewed (approved):
Merge at will

Commit:
Remove Non-ASCII character in transformer.py.
